File: BlogModule.Application/QueryHandlers/BlogPost/GetBlogPost/GetBlogPostQueryHandler.cs
using System.Collections.Immutable;
using System.Text.RegularExpressions;
using BlogModule.Application.Contract.BlogPosts.GetBlogPost;
using BlogModule.Domain.Models;
using BlogModule.Domain.Models.Ids;
using BlogModule.Domain.RepositoryInterfaces.BlogPost;
using Results.Application;
using Results.Domain;
using Shared.CQRS;
using Shared.CQRS.QueryHandlers;

namespace BlogModule.Application.QueryHandlers.BlogPost.GetBlogPost;

public class GetBlogPostQueryHandler : IQueryHandler<GetBlogPostQuery, HandlerResult<GetBlogPostResponse, IErrorResult>>
{
    private readonly IBlogPostQueryRepository _repository;

    public GetBlogPostQueryHandler(IBlogPostQueryRepository repository)
    {
        _repository = repository;
    }

    public async Task<HandlerResult<GetBlogPostResponse, IErrorResult>> Handle(GetBlogPostQuery request,
        CancellationToken cancellationToken)
    {
        var isSuccess = Guid.TryParse(request.IdOrUrl, out var blogPostId);

        var blogPost = isSuccess
            ? await _repository.GetByIdAsync(new BlogPostId(blogPostId), cancellationToken)
            : await _repository.GetByUrlAsync(request.IdOrUrl, cancellationToken);


        return blogPost is null
            ? new EntityNotFoundErrorResult()
            {
                Title = "Blog post was not found",
                Message = $"Blog post with ID/Url {request.IdOrUrl} was not found in the database."
            }
            : new GetBlogPostResponse
            {
                Id = blogPost.Id,
                Heading = blogPost.Heading,
                PageTitle = blogPost.PageTitle,
                Author = blogPost.Author,
                Visible = blogPost.Visible,
                Content = blogPost.Content,
                ShortDescription = blogPost.ShortDescription,
                FeaturedImageUrl = blogPost.FeaturedImageUrl,
                UrlHandle = blogPost.UrlHandle,
                PublishedDate = blogPost.PublishedDate,
                Tags = blogPost.Tags.ToImmutableArray(),
                Likes = blogPost.Likes.ToImmutableArray(),
                Comments = blogPost.Comments.ToImmutableArray(),
            };
    }
}

File: BlogApi/Application/query_handlers/blog_post/get_blog_post/get_blog_post_query_handler.go
package queries

import (
	contract "blog-api/Application.contract/BlogPosts/GetBlogPost"
	repository_interfaces "blog-api/Domain/repository_interfaces/blog_post_repository"
	"blog-api/Domain/responses"
	"context"
)

type GetBlogPostQueryHandler struct {
	blogPostQueryRepository repository_interfaces.BlogPostQueryRepository
}

func NewGetBlogPostQueryHandler(
	blogPostQueryRepository repository_interfaces.BlogPostQueryRepository,
) *GetBlogPostQueryHandler {
	return &GetBlogPostQueryHandler{
		blogPostQueryRepository: blogPostQueryRepository,
	}
}

func (h *GetBlogPostQueryHandler) Handle(ctx context.Context, query *GetBlogPostQuery) (*contract.GetBlogPostResponse, error) {
	blogPost, err := h.blogPostQueryRepository.GetBlogPostByID(query.ID)
	if err != nil {
		response := responses.NewResponse[contract.GetBlogPostResponse](500, "Failed to retrieve blog post")
		return &response, nil
	}

	if blogPost == nil {
		response := responses.NewResponse[contract.GetBlogPostResponse](404, "Blog post not found")
		return &response, nil
	}

	return &contract.GetBlogPostResponse{
		BaseResponse: responses.NewBaseResponse(200, "Blog post retrieved successfully"),
		BlogPost:     *blogPost,
	}, nil
}
